libs/server-sdk/src/evaluation/detail/semver_operations.cpp

```diff
@@ -0,0 +1,203 @@
+
+/**
+ * Cache the parsed regex so it doesn't need to be rebuilt constantly.
+ *
+ * From Boost docs:
+ * Class basic_regex and its typedefs regex and wregex are thread safe, in that
+ * compiled regular expressions can safely be shared between threads.
+ */
+static boost::regex const& SemVerRegex() {
+    static boost::regex regex{kSemVerRegex, boost::regex_constants::no_except};
+    LD_ASSERT(regex.status() == 0);
+    return regex;
+}
+
+SemVer::SemVer(VersionType major,
+               VersionType minor,
+               VersionType patch,
+               std::vector<Token> prerelease)
+    : major_(major), minor_(minor), patch_(patch), prerelease_(prerelease) {}
+
+SemVer::SemVer(VersionType major, VersionType minor, VersionType patch)
+    : major_(major), minor_(minor), patch_(patch), prerelease_(std::nullopt) {}
+
+SemVer::SemVer() : SemVer(0, 0, 0) {}
+
+SemVer::VersionType SemVer::Major() const {
+    return major_;
+}
+
+SemVer::VersionType SemVer::Minor() const {
+    return minor_;
+}
+
+SemVer::VersionType SemVer::Patch() const {
+    return patch_;
+}
+
+std::optional<std::vector<SemVer::Token>> const& SemVer::Prerelease() const {
+    return prerelease_;
+}
+
+bool operator<(SemVer::Token const& lhs, SemVer::Token const& rhs) {
+    if (lhs.index() != rhs.index()) {
+        /* Numeric identifiers (index 0 of variant) always have lower precedence
+than non-numeric identifiers. */
+        return lhs.index() < rhs.index();
+    }
+    if (lhs.index() == 0) {
+        return std::get<0>(lhs) < std::get<0>(rhs);
+    }
+    return std::get<1>(lhs) < std::get<1>(rhs);
+}
+
+bool operator==(SemVer const& lhs, SemVer const& rhs) {
+    return lhs.Major() == rhs.Major() && lhs.Minor() == rhs.Minor() &&
+           lhs.Patch() == rhs.Patch() && lhs.Prerelease() == rhs.Prerelease();
+}
+
+bool operator<(SemVer const& lhs, SemVer const& rhs) {
+    if (lhs.Major() < rhs.Major()) {
+        return true;
+    }
+    if (lhs.Major() > rhs.Major()) {
+        return false;
+    }
+    if (lhs.Minor() < rhs.Minor()) {
+        return true;
+    }
+    if (lhs.Minor() > rhs.Minor()) {
+        return false;
+    }
+    if (lhs.Patch() < rhs.Patch()) {
+        return true;
+    }
+    if (lhs.Patch() > rhs.Patch()) {
+        return false;
+    }
+    // At this point, lhs and rhs have equal major/minor/patch versions.
+    if (!lhs.Prerelease() && !rhs.Prerelease()) {
+        return false;
+    }
+    if (lhs.Prerelease() && !rhs.Prerelease()) {
+        return true;
+    }
+    if (!lhs.Prerelease() && rhs.Prerelease()) {
+        return false;
+    }
+    return *lhs.Prerelease() < *rhs.Prerelease();
+}
+
+bool operator>(SemVer const& lhs, SemVer const& rhs) {
+    return rhs < lhs;
+}
+
+std::optional<SemVer> SemVer::Parse(std::string const& value) {
```

Comment:
using `boost::lexical_cast` instead of `std::stoull` because the latter accepts things like "2-3" as the number 2.